sql/src/main/java/io/crate/operation/udf/UserDefinedFunctionService.java

```diff
@@ -21,44 +21,110 @@
 
-import io.crate.metadata.FunctionIdent;
-import org.elasticsearch.cluster.ClusterService;
+import com.google.common.annotations.VisibleForTesting;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.*;
+import org.elasticsearch.cluster.ack.ClusterStateUpdateRequest;
+import org.elasticsearch.cluster.ack.ClusterStateUpdateResponse;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.common.Nullable;
 import org.elasticsearch.common.component.AbstractComponent;
 import org.elasticsearch.common.inject.Inject;
 import org.elasticsearch.common.settings.Settings;
 
-public class UserDefinedFunctionService extends AbstractComponent {
+import java.util.ArrayList;
+import java.util.List;
+
+public class UserDefinedFunctionService extends AbstractComponent implements ClusterStateListener {
 
     private final ClusterService clusterService;
 
     @Inject
     public UserDefinedFunctionService(Settings settings, ClusterService clusterService) {
         super(settings);
         this.clusterService = clusterService;
+        clusterService.add(this);
     }
 
     /**
-     * Returns registered function
-     *
-     * @param functionIdent
-     * @return
+     * This method can only be called on master node
+     * registers new function in the cluster
+     * @param request function to register
      */
-    public UserDefinedFunction function(FunctionIdent functionIdent) {
-        return null;
+    void registerFunction(final RegisterUserDefinedFunctionRequest request, final ActionListener<ClusterStateUpdateResponse> listener) {
+        if (!clusterService.localNode().isMasterNode()) {
+            return;
+        }
+        clusterService.submitStateUpdateTask(request.cause, new AckedClusterStateUpdateTask<ClusterStateUpdateResponse>(request, listener) {
+
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                MetaData metaData = currentState.metaData();
+                MetaData.Builder mdBuilder = MetaData.builder(currentState.metaData());
+                UserDefinedFunctionsMetaData functions = putFunction(metaData.custom(UserDefinedFunctionsMetaData.TYPE), request.metaData);
+                mdBuilder.putCustom(UserDefinedFunctionsMetaData.TYPE, functions);
+                return ClusterState.builder(currentState).metaData(mdBuilder).build();
+            }
+
+            @Override
+            protected ClusterStateUpdateResponse newResponse(boolean acknowledged) {
+                return new ClusterStateUpdateResponse(acknowledged);
+            }
+
+        });
     }
 
-    /**
-     * registers new function in the cluster
-     * @param function function to register
-     */
-    public void registerFunction(UserDefinedFunction function) {
 
+    @VisibleForTesting
+    static UserDefinedFunctionsMetaData putFunction(@Nullable UserDefinedFunctionsMetaData functions, UserDefinedFunctionMetaData function) {
```

Comment:
i would not allow it to be nullable, then we do need that null and will just loop over functions if it has smth